streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -68,72 +69,115 @@ private static synchronized void updateNumRecordsProduces(final int delta) {
+            producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, "EosTest");
+            producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+            producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+            producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+            producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
 
-        final Random rand = new Random(System.currentTimeMillis());
+            final KafkaProducer<String, Integer> producer = new KafkaProducer<>(producerProps);
 
-        while (isRunning) {
-            final String key = "" + rand.nextInt(MAX_NUMBER_OF_KEYS);
-            final int value = rand.nextInt(10000);
+            final Random rand = new Random(System.currentTimeMillis());
+            final Map<Integer, List<Long>> offsets = new HashMap<>();
 
-            final ProducerRecord<String, Integer> record = new ProducerRecord<>("data", key, value);
+            while (isRunning) {
+                final String key = "" + rand.nextInt(MAX_NUMBER_OF_KEYS);
+                final int value = rand.nextInt(10000);
 
-            producer.send(record, (metadata, exception) -> {
-                if (exception != null) {
-                    exception.printStackTrace(System.err);
-                    System.err.flush();
-                    if (exception instanceof TimeoutException) {
-                        try {
-                            // message == org.apache.kafka.common.errors.TimeoutException: Expiring 4 record(s) for data-0: 30004 ms has passed since last attempt plus backoff time
-                            final int expired = Integer.parseInt(exception.getMessage().split(" ")[2]);
-                            updateNumRecordsProduces(-expired);
-                        } catch (final Exception ignore) { }
+                final ProducerRecord<String, Integer> record = new ProducerRecord<>("data", key, value);
+
+                producer.send(record, (metadata, exception) -> {
+                    if (exception != null) {
+                        exception.printStackTrace(System.err);
+                        System.err.flush();
+                        if (exception instanceof TimeoutException) {
+                            try {
+                                // message == org.apache.kafka.common.errors.TimeoutException: Expiring 4 record(s) for data-0: 30004 ms has passed since last attempt plus backoff time
+                                final int expired = Integer.parseInt(exception.getMessage().split(" ")[2]);
+                                updateNumRecordsProduces(-expired);
+                            } catch (final Exception ignore) {
+                            }
+                        }
+                    } else {
+                        offsets.getOrDefault(metadata.partition(), new LinkedList<>()).add(metadata.offset());
```

Comment:
How does this logic work? Are we updating the offsets map when the partition is not present?

Reply:
If the partition is not present, we first create a new entry in the map for the partition with an empty list. Afterward we add to the empty list. Ie, the code is short for:
```
if (!offsets.containsKey(metadata.partition()) {
  offsets.put(metadata.partition(), new LinkedList<>());
}
offsets.get(metadata.partition()).add(metadata.offset());
```